js/src/elements/star-rating.ts

```diff
@@ -0,0 +1,355 @@
+		<label for="radio-3">
+			<svg aria-hidden="true" xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" style="overflow: visible" >
+				<use href="#star">
+			</svg>
+			<span part="visually-hidden">3<!-- default label, will be replaced by slot content if provided --></span>
+		</label>
+
+		<input part="visually-hidden" type="radio" value="4" id="radio-4" />
+		<label for="radio-4" >
+			<svg aria-hidden="true" xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" style="overflow: visible" >
+				<use href="#star">
+			</svg>
+			<span part="visually-hidden">4<!-- default label, will be replaced by slot content if provided --></span>
+		</label>
+
+		<input part="visually-hidden" type="radio" value="5" id="radio-5" />
+		<label for="radio-5">
+			<svg aria-hidden="true" xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" style="overflow: visible" >
+				<use href="#star">
+			</svg>
+			<span part="visually-hidden">5<!-- default label, will be replaced by slot content if provided --></span>
+		</label>
+
+		<output id="alert" for="radio-1 radio-2 radio-3 radio-4 radio-5">
+			<span id="label-1"><slot name="label-1"></slot></span>
+			<span id="label-2"><slot name="label-2"></slot></span>
+			<span id="label-3"><slot name="label-3"></slot></span>
+			<span id="label-4"><slot name="label-4"></slot></span>
+			<span id="label-5"><slot name="label-5"></slot></span>
+		</output>
+	</div>
+    </fieldset>
+`;
+
+const template = starRatingTemplate;
+
+class StarRatingElement extends HTMLElement {
+	static get observedAttributes() {
+		return ['disabled', 'name', 'required', 'value'];
+	}
+
+	coercedValue = '';
+
+	constructor() {
+		super();
+		this.attachShadow({ mode: 'open' });
+		this.shadowRoot?.appendChild(template.content.cloneNode(true));
+	}
+
+	get disabled() {
+		return this.hasAttribute('disabled');
+	}
+
+	set disabled(val) {
+		this.setAttribute('disabled', val.toString());
+	}
+
+	get form() {
+		return this.closest(`form`);
+	}
+
+	get name() {
+		return this.getAttribute('name') ?? '';
+	}
+
+	set name(val) {
+		this.setAttribute('name', val);
+	}
+
+	get required() {
+		return this.hasAttribute('required');
+	}
+
+	set required(val) {
+		this.required = val !== null;
+	}
+
+	get type() {
+		return this.localName;
+	}
+
+	get value() {
+		return this.coercedValue;
+	}
+
+	set value(val: string) {
+		this.coercedValue = ['', '1', '2', '3', '4', '5'].includes(val) ? val : '';
+		const toCheck = this.shadowRoot?.querySelector(
+			`[value="${this.coercedValue}"]`
+		) as HTMLInputElement;
+		if (toCheck) {
+			toCheck.checked = true;
+		} else {
+			const uncheck = this.shadowRoot?.querySelector(':checked') as HTMLInputElement;
+			if (uncheck) {
+				uncheck.checked = false;
+			}
+		}
+	}
+
+	get valueAsNumber() {
+		return parseInt(this.getAttribute('value') ?? '');
+	}
+
+	get validity() {
+		// borrow input's validity, https://caniuse.com/?search=validity
```

Comment:
Internals is not supported for usage yet, so can't hook into internals/form control API. `Validity` refers to a read-only `ValidityState` interface. Rather than trying to create our own `ValidityState` interface, it may be simpler to just sync with the active input's validity, or the first input if not active.

https://web.dev/more-capable-form-controls/
https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity